man/man1/wchars.1
.TH WCHARS 1 "Version 1, Release 4" "W Window System" "W PROGRAMS"
.SH NAME
wchars \- show characters in the given W font
.SH SYNOPSIS
.nf
wchars <font family> <font size> [<styles: bilru>]
.fi
.SH DESCRIPTION
.I Wchars
shows the whole character table for the font that Wlib returns with
given parameters.  If you give "" as the family, or 0 as size, the
server will use the default value.  Styles that can be requested with
identification letters are:  bold, italics, light, reverse, underline.
.PP
If you left-click a character or press a key in the window, character in
question is inverted and it's ASCII-code is shown in the window title.
.PP
You can insert the character into Wlib clipboard with a right-click.
.SH AUTHOR
(w) 11/1996 by Eero Tamminen


wt/CHANGES
++oddie 9/2000:
- fixed listbox/filesel resizing & added minimum size setopt.

++bsittler 7/2000:
- Fixed vt color/reverse and attribute resetting
- Add HTTP/1.1 'Host' header to Wetscape requests
- Other minor changes (netgames, wvt, waudio)

++jcarlson 7/2000:
- WTINYSCREEN changes

++eero, 5/1999:
- Fixed key handling for special keys in Wt/getstring.c & wedit/wedit.c.

++eero, 1/1999:
- Added support for new W special keys so that now on W platforms which
implement special key mappings, you don't need to use Emacs shortcuts
with text widgets (or remap your console cursor keys)...


------ notes for version released with W R1PL0 ----------

ADDITIONAL WIDGETS

- Arrow, a simple button with an arrow drawn into it.

- Dial, a dial widget where you can position a needle into desired
  angle.  This isn't in the W toolkit library itself as then you would
  need to link math library to all of your Wt programs.  You'll find it
  in the warzone game tar-ball.

- Edittext, mainly intended for simple one page stuff, but as this
  spaghetti can do some additional stuff like paragraph formatting,
  there's a real abuse built on this, the WEdit.  As edittext widget is
  large and buggy, it's in the WEdit directory instead of the toolkit.

- Menu, a strip of buttons from which popups pop-up.  Both the popups
  and menus use a structure defined in Wt.h instead of strings that Kay
  was using.  The reason is that I want callbacks to follow
  automatically hierarchically opened popups.

- Popup, hierarchical popups.  Can include subpopups, images, lines and
  in addition items can also be disabled and checked.

- Range, a scrollbar composite with which you can select a value between
  given limits.

- Select, a getstring widget coupled with an arrow openable listbox from
  which items can be selected into a getstring widget.


OTHER CHANGES

- vt can now do everything wterm  can do + user can rebind the colors
  in wtrc.

- Added '.wtconfig' file and made makefiles use the W compilation
  configuration file '.config' in parent directory.

- wt_do_event() doesn't call wt_dispatch_event() anymore with
  EVENT_GADGET/GADGET_EXIT event as those might not have valid window
  pointer.  If you want to check for abnormal (unboud) exit, check the
  wt_return value.  GADGET_EXIT returns a negative number and
  GADGET_CLOSE (in shell widget) a positive number now.

- Added xterm mouse events, reverse mode, fd/bgcolor setting + improved
  double/triple click timing in VT widget.

- Aliases a couple new tags into others in HTML widget and added
  HTML header (ignored) and TITLE parsing.

- Widget event processing functions return now unrecognized or
  non-processed WEVENTs or NULL instead of long.  Assumption is that if
  function returns WEVENT, the widget whose action it was, is still
  around.  If in doubt, use NULL.  Unrecognized key events should be
  returned for focus change though.  (actually it could still have been
  an int, but returning a WEVENT offers a possibility for event mapping
  widgets.)

- Added wt_global structure containing Wt configuration variables and
  made widgets to use it.  wt_init() will now read these variables from
  ~/.wtrc + a list of (variable, value) pairs.  Widget specific
  initialization code in their init() functions may read these with the
  wt_variable() function.

- HTML widget fonts can be configured by above method.

- Added wt_loadfont() function which will use the user specified Wt
  default if widget doesn't specify some font parameter.  It also exits
  program if Wt default fonts can't be loaded, because otherwise
  unavailable fonts may chrash naive programs.

- Changed widgets to reflect the new w_loadfont() syntax (NULL/zero
  values tell Wlib/server to use server defaults).

- Added new WT_FONTSIZE define.  The earlier WT_FONT setopt sets now the
  font *family* name, new one sets the font size.

- Getopting WT_FONT returns WFONT* pointer also for above reason.

- Removed 'fontname' members from widget structures.  Needed information
  and more can be gotten from the 'font' member if needed.

- Fileselector has now separate lists for directory and file names.
  Both of course have an autolocator (type & return in addition to point
  & click).

- Wt_keys.h defines values for some special keys like cursors etc.  Half
  of the defines are Emacs style and half what the special keys itself
  give or are in use on other platforms.

- Changed Wt option defines.  There were a couple overlapping ones.


------- Kay's original notes: ----------------------------------------------

- added some missing close() and truncate the file to zero length before
  writing a new selection in selection.c.

- better selection handling in VT widget.

- made everything resizeable.

- applied patches by eero and fixed some more bugs myself.

- implemented a simple clipboard and changed getstring and vt widget to use it.

- fixed several serious bugs in the html widget.

- wrote VT widget.

- added an option to the icon editor to sync edit field with display
  field (simon).

- added missing special chars to html widget (simon).

- added code to the fileselector to set a default file name.

- lots of changes by eero (button.c, fileselect.c, listbox.c, getstring.c), some
  fixes to his changes by me.

- added missing wt_ungetfocus() to listbox and getstring.

- wrote Tk-like packer widget.

- new fileselect widget and testprogram by eero. Added better filename
  matching (eeros version was a joke, really).

- getstring and listbox updates by eero.

- changed the scrollbar widget's user interface to become more user friendly.
  See scrollbar.doc for details on how to use it. Also changed viewport and
  listbox to use the new interface.

- added listbox_focus(). Otherwise listbox cannot receive keyboard focus...

- big bugfix: when wt_delete() is called on a widget all pending timers must
  be canceled! (affected list, iconedit and getstring).


apps/wchars.c
/*
 * wchars.c, a part of the W Window System
 *
 * Copyright (C) 1997-1998 by Eero Tamminen
 *
 * This package is free software; you can redistribute it and/or modify it
 * under the terms specified in the docs/COPYRIGHTS file coming with this
 * package.
 *
 * -- show characters in the given W font
 */

#include <stdio.h>
#include <stdlib.h>
#include <Wlib.h>

static short xsize, ysize;		/* (maximum) char size */

#define WIN_FLAGS (W_MOVE | W_TITLE | W_CLOSE | EV_KEYS | EV_MOUSE)


static void show_key(WWIN *win, WFONT *font, int key)
{
  char msg[16];
  int x, y;

  if(key < 0 || key > 255)
    return;

  x = key % 32;
  y = key / 32;
  x = xsize * x + x + (xsize - font->widths[y * 32 + x]) / 2,
  y = ysize * y + y;

  w_pbox(win, x, y, font->widths[key], ysize);
  sprintf(msg, " char: %03d ", key);
  w_settitle(win, msg);
}

int main(int argc, char *argv[])
{
  int x, y, size, old = -1;
  ushort styles;
  char *family;
  WFONT *font;
  WEVENT *ev;
  WWIN *win;
  uchar msg;
 
  if((argc != 3 && argc != 4) || argv[1][0] == '-')
  {
    fprintf(stderr, "Show font character table\n");
    fprintf(stderr, "usage: %s <family> <size> [<styles: bilru>]\n", *argv);
    fprintf(stderr, "for example: %s fixed 14 b\n", *argv);
    return -1;
  }

  family = argv[1];
  size = atoi(argv[2]);

  styles = 0;
  if (argc == 4)
  {
    int idx = 0;
    for (;;)
    {
      switch(argv[3][idx++])
      {
	case 'b':
	  styles |= F_BOLD;
	  break;
	case 'i':
	  styles |= F_ITALIC;
	  break;
	case 'l':
	  styles |= F_LIGHT;
	  break;
	case 'r':
	  styles |= F_REVERSE;
	  break;
	case 'u':
	  styles |= F_UNDERLINE;
	  break;
	default:
	  /* sure, I'd like a multilevel break, but this is OK too... */
	  goto styles_done;
      }
    }
  }
styles_done:


  if(!w_init())
    return -1;

  if(!(font = w_loadfont(family, size, styles)))
  {
    fprintf(stderr, "%s: unable to load font %s\n", argv[0], argv[1]);
    return -1;
  }
  xsize = font->maxwidth;
  ysize = font->height;

  if (!(win = w_create(xsize * 32 + 31, ysize * 8 + 7, WIN_FLAGS)))
  {
    fprintf(stderr, "%s: unable to create an output window\n", *argv);
    return -1;
  }
  w_setmode(win, M_DRAW);
  w_setfont(win, font);

  /* character grid */
  for(y = 1; y < 8; y++)
    w_hline(win, 0, ysize * y + y - 1, win->width);
  for(x = 1; x < 32; x++)
    w_vline(win, xsize * x + x - 1, 0, win->height);

  /* characters */
  for(y = 0; y < 8; y++)
    for(x = 0; x < 32; x++)
      w_printchar(win,
        xsize * x + x + (xsize - font->widths[y * 32 + x]) / 2,
	ysize * y + y,
	32 * y + x);

  w_settitle(win, " ASCII-table ");
  if (w_open(win, UNDEF, UNDEF) < 0)
  {
    fprintf(stderr, "%s: unable to open an output window\n", *argv);
    w_delete(win);
    return -1;
  }

  w_setmode(win, M_INVERS);
  for(;;)
  {
    if((ev = w_queryevent(NULL, NULL, NULL, -1)))
    {
      switch(ev->type)
      {
        case EVENT_KEY:
	  show_key(win, font, old);		/* remove old mark */
	  show_key(win, font, ev->key);
	  old = ev->key;
	  break;

	case EVENT_MPRESS:
	  show_key(win, font, old);		/* remove old mark */
	  break;

	case EVENT_MRELEASE:
	  old = (ev->y / (ysize + 1)) * 32 + ev->x / (xsize + 1);
	  show_key(win, font, old); 
	  if(ev->key == BUTTON_RIGHT)
	  {
	    msg = old;
	    w_putselection(W_SEL_TEXT, &msg, 1);
	  }
	  break;

	case EVENT_GADGET:
	if(ev->key == GADGET_EXIT || ev->key == GADGET_CLOSE)
	{
	  w_unloadfont(font);
	  w_delete(win);
	  return 0;
	}
      }
    }
  }
  return 0;
}
